### DISCUSSION TITLE
Recommended way to set public ip on non-cloud or bare metal nodes?

### DISCUSSION THREAD

**Discussion by justyns** [siderolabs/talos#8923] [Q&A]
I am testing Talos+Omni+Kubespan out and one question I haven't figured out yet is how do you appropriately set the external/public ip for a node?

For example, I have 5 nodes inside a private network and no direct public ip.  These show up correctly.   I also have one more node that is running on a public vps that only has a single ipv4, no private/internal ips at all.   This shows the public ip as the internal ip and nothing for the public ip.

```
❯ kubectl get nodes -o wide
NAME             STATUS   ROLES           AGE     VERSION   INTERNAL-IP       EXTERNAL-IP   OS-IMAGE         KERNEL-VERSION   CONTAINER-RUNTIME
external-test1   Ready    <none>          3h47m   v1.30.1   23.1.2.3       <none>        Talos (v1.7.4)   6.6.32-talos     containerd://1.7.16
talos-3lv-a9l    Ready    <none>          21h     v1.30.1   192.168.226.116   <none>        Talos (v1.7.4)   6.6.32-talos     containerd://1.7.16
talos-mfq-cq2    Ready    control-plane   21h     v1.30.1   192.168.226.115   <none>        Talos (v1.7.4)   6.6.32-talos     containerd://1.7.16
talos-olj-ijd    Ready    control-plane   21h     v1.30.1   192.168.226.114   <none>        Talos (v1.7.4)   6.6.32-talos     containerd://1.7.16
talos-qfz-a4u    Ready    control-plane   21h     v1.30.1   192.168.226.113   <none>        Talos (v1.7.4)   6.6.32-talos     containerd://1.7.16
talos-rap-34g    Ready    <none>          21h     v1.30.1   192.168.226.117   <none>        Talos (v1.7.4)   6.6.32-talos     containerd://1.7.16
```

I saw a talos FAQ which suggested it's the responsibility of the cloud controller manager.  The VPS I'm using does not have any sort of cloud metadata service/etc.  I also have several bare metal servers that will be in a similar situation.  I installed https://github.com/siderolabs/talos-cloud-controller-manager thinking it may help, but wasn't sure what to actually configure with it.  The public ip didn't seem to be editable.

Also with kubespan, should the internal ip end up being the kubespan ip for the node?

**Comment by smira** [comment]
It is correct that it's no the OS responsibility to set the external IP field for the Kubernetes `Node` resource, and Talos Linux itself won't do that.

You can set it manually or by implementing a controller manager of some sorts, but Talos Linux itself doesn't know about public/private IPs except for specific platforms (e.g. AWS, GCP, ...). 

KubeSpan acts on machine IPs, so it doesn't care about Kubernetes `Node` resources, basically it acts on whatever appears in `talosctl get nodeaddresses` (but ignoring Kubernetes pod/service addresses by default), this is not directly related to the Kubernetes view of things.

**Comment by kvaps** [comment]
Hey, just for your information: I ran into the same issue and created local-ccm, which automatically detects the external address via netlink and sets it on the Node object.

An example configuration for Talos is provided here:
https://github.com/cozystack/local-ccm?tab=readme-ov-file#talos-linux
